### PR TITLE
Enable anti join type in SMJ 

### PR DESCRIPTION
**Pull request by JkSelf**
Anti join is equal to the left join with the left side rows that have no match on the right.

### CODE REVIEW

File: velox/exec/fuzzer/JoinFuzzer.cpp [hunk 1]
```diff
@@ -963,7 +963,8 @@ void JoinFuzzer::makeAlternativePlans(
           .planNode()});
 
   // Use OrderBy + MergeJoin (if join type is inner or left).
```
Comment (jinchengchenghh) [inline velox/exec/fuzzer/JoinFuzzer.cpp:965]: Update the comment

Comment (JkSelf) [reply velox/exec/fuzzer/JoinFuzzer.cpp]: @jinchengchenghh Updated.

**Comment by JkSelf** [comment]
@pedroerp @mbasmanova @rui-mo @jinchengchenghh Can you help to review this PR? Thanks.

File: velox/exec/MergeJoin.cpp [hunk 2]
```diff
@@ -89,10 +89,16 @@ void MergeJoin::initialize() {
   if (joinNode_->filter()) {
     initializeFilter(joinNode_->filter(), leftType, rightType);
 
-    if (joinNode_->isLeftJoin()) {
+    if (joinNode_->isLeftJoin() || joinNode_->isAntiJoin()) {
       leftJoinTracker_ = LeftJoinTracker(outputBatchSize_, pool());
     }
   }
+
+  // Anti join need to track the left side rows that have no match on the right.
```
Comment (rui-mo) [inline velox/exec/MergeJoin.cpp:97]: nit: need -> needs.

File: velox/exec/MergeJoin.cpp [hunk 3]
```diff
@@ -321,6 +327,12 @@ void MergeJoin::addOutputRow(
     }
   }
 
+  // Anti join need to track the left side rows that have no match on the right.
```
Comment (rui-mo) [inline velox/exec/MergeJoin.cpp:330]: ditto.

File: velox/exec/MergeJoin.cpp [hunk 4]
```diff
@@ -565,7 +577,34 @@ RowVectorPtr MergeJoin::getOutput() {
...
+
+          if (numPassed == numRows) {
+            // All rows passed.
+            return output;
+          }
+
+          // Some, but not all rows passed.
+          return wrap(numPassed, indices, output);
+        } else {
+          return output;
```
Comment (rui-mo) [inline velox/exec/MergeJoin.cpp:606]: I wonder if it's better to make let this branch return first to improve the readability
```
If (!isAntiJoin(joinType_)) {
  return output;
}
auto numRows = output->size();
...
```

**Review by rui-mo — COMMENTED** [review]
Thanks! I wonder if we have noticed the documentation for the null semantics of anti join. Perhaps we could check against it to ensure the implementation matches.

https://facebookincubator.github.io/velox/develop/anti-join.html#anti-join

File: velox/exec/MergeJoin.cpp [hunk 5]
```diff
@@ -565,7 +577,34 @@ RowVectorPtr MergeJoin::getOutput() {
         // No rows survived the filter. Get more rows.
         continue;
       } else {
```
Comment (jinchengchenghh) [inline velox/exec/MergeJoin.cpp:579]: else if (isAntiJoin(joinType_))

File: velox/exec/fuzzer/JoinFuzzer.cpp [hunk 6]
```diff
@@ -953,9 +953,11 @@ void JoinFuzzer::makeAlternativePlans(
               joinNode->isNullAware())
           .planNode()});
 
-  // Use OrderBy + MergeJoin (if join type is inner or left).
+  // Use OrderBy + MergeJoin (if join type is inner, left, left semi, right semi
```
Comment (jinchengchenghh) [inline velox/exec/fuzzer/JoinFuzzer.cpp:956]: I think we can drop the if in blank, code says all

File: velox/exec/MergeJoin.cpp (hunk 5 above)
Comment (JkSelf) [reply velox/exec/MergeJoin.cpp]: @jinchengchenghh Updated.

**Comment by JkSelf** [comment]
> Thanks! I wonder if we have noticed the documentation for the null semantics of anti join. Perhaps we could check against it to ensure the implementation matches.
> 
> https://facebookincubator.github.io/velox/develop/anti-join.html#anti-join

@rui-mo It seems only [hash join](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastHashJoinExec.scala#L48) need to consider null aware and [smj ](https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/SortMergeJoinExec.scala#L39)not need to consider. Correct me if wrong understand. Thanks.

**Review by pedroerp — COMMENTED** [review]
Never mind, I missed you already added fuzzer support!

File: velox/exec/MergeJoin.cpp [hunk 7]
```diff
@@ -89,10 +89,17 @@ void MergeJoin::initialize() {
   if (joinNode_->filter()) {
     initializeFilter(joinNode_->filter(), leftType, rightType);
 
-    if (joinNode_->isLeftJoin()) {
+    if (joinNode_->isLeftJoin() || joinNode_->isAntiJoin()) {
       leftJoinTracker_ = LeftJoinTracker(outputBatchSize_, pool());
     }
   }
+
```
Comment (pedroerp) [inline velox/exec/MergeJoin.cpp:96]: should this be a 

```
} else if (joinNode_->isAntiJoin()) {
```

?

File: velox/exec/MergeJoin.cpp [hunk 8]
```diff
@@ -321,6 +328,13 @@ void MergeJoin::addOutputRow(
     }
   }
 
+  // Anti join needs to track the left side rows that have no match on the
+  // right.
+  if (isAntiJoin(joinType_) && !filter_ && leftJoinTracker_) {
```
Comment (pedroerp) [inline velox/exec/MergeJoin.cpp:333]: should we here just assert that `leftJoinTracker_` exists instead?
```
if (isAntiJoin(joinType_)) {
  VELOX_CHECK(leftJoinTracker_);
```

File: velox/exec/MergeJoin.cpp [hunk 9]
```diff
@@ -564,6 +578,31 @@ RowVectorPtr MergeJoin::getOutput() {
 
         // No rows survived the filter. Get more rows.
         continue;
+      } else if (isAntiJoin(joinType_)) {
```
Comment (pedroerp) [inline velox/exec/MergeJoin.cpp:581]: could we move this logic to a helper function, and comment what you are doing? This function is already pretty long and hard to follow.

File: velox/exec/MergeJoin.cpp (hunk 7 above)
Comment (JkSelf) [reply velox/exec/MergeJoin.cpp]: @pedroerp Yes. Updated.

File: velox/exec/MergeJoin.cpp (hunk 8 above)
Comment (JkSelf) [reply velox/exec/MergeJoin.cpp]: @pedroerp Updated.

File: velox/exec/MergeJoin.cpp (hunk 9 above)
Comment (JkSelf) [reply velox/exec/MergeJoin.cpp:607]: @pedroerp Yes. Added filterOutputForAntiJoin method.

**Comment by JkSelf** [comment]
@pedroerp I have resolved all your comments. Can you help to review again? Thanks.

**Review by pedroerp — APPROVED** [review]
Thank you @JkSelf, looks good to me.
